File: src/cpp/surface-generation/biomes/noise/normal-noise.cpp
#include "normal-noise.hpp"

double expectedDeviation(int32_t v) {
    return 0.1 * (1.0 + 1.0 / (double)(v + 1));
}

// NormalNoise

NormalNoise NormalNoise::createLegacyNetherBiome(RandomSource *randomSource,
                                                 NormalNoise::NoiseParameters const &parameters) {
    return NormalNoise(randomSource, parameters.firstOctave, parameters.amplitudes, false);
}

NormalNoise NormalNoise::create(RandomSource *randomSource, NormalNoise::NoiseParameters const &parameters) {
    return NormalNoise(randomSource, parameters.firstOctave, parameters.amplitudes, true);
}

NormalNoise NormalNoise::create(RandomSource *randomSource, int32_t firstOctave, vector<double> const &amplitudes) {
    return NormalNoise(randomSource, firstOctave, amplitudes, true);
}

NormalNoise::NormalNoise(RandomSource *randomSource, int32_t firstOctave, vector<double> const &amplitudes,
                         bool notLegacy) {
    if (notLegacy) {
        this->first = PerlinNoise::create(randomSource, firstOctave, amplitudes);
        this->second = PerlinNoise::create(randomSource, firstOctave, amplitudes);
    } else {
        this->first = PerlinNoise::createLegacyForLegacyNormalNoise(randomSource, firstOctave, amplitudes);
        this->second = PerlinNoise::createLegacyForLegacyNormalNoise(randomSource, firstOctave, amplitudes);
    }

    int32_t minAmplitudeIndex = numeric_limits<int32_t>::max();
    int32_t maxAmplitudeIndex = numeric_limits<int32_t>::lowest();

    for (int32_t amplitudeIndex = 0; amplitudeIndex < amplitudes.size(); amplitudeIndex++) {
        double amplitude = amplitudes.at(amplitudeIndex);
        if (amplitude != 0.0) {
            minAmplitudeIndex = min(minAmplitudeIndex, amplitudeIndex);
            maxAmplitudeIndex = max(maxAmplitudeIndex, amplitudeIndex);
        }
    }

    this->valueFactor = 0.16666666666666666 / expectedDeviation(maxAmplitudeIndex - minAmplitudeIndex);
}

double NormalNoise::getValue(double x, double y, double z) const {
    double inputX = x * INPUT_FACTOR;
    double inputY = y * INPUT_FACTOR;
    double inputZ = z * INPUT_FACTOR;
    return (this->first->getValue(x, y, z) + this->second->getValue(inputX, inputY, inputZ)) * this->valueFactor;
}

// NoiseParameters

NormalNoise::NoiseParameters::NoiseParameters(int32_t firstOctave, vector<double> const &amplitudes) {
    this->firstOctave = firstOctave;
    this->amplitudes = amplitudes;
}

NormalNoise::NoiseParameters::NoiseParameters(int32_t firstOctave, double firstAmplitude,
                                              vector<double> const &amplitudes) {
    this->firstOctave = firstOctave;
    this->amplitudes = vector<double>();
    this->amplitudes.reserve(1 + amplitudes.size());
    this->amplitudes.insert(this->amplitudes.begin(), firstAmplitude);
    this->amplitudes.insert(this->amplitudes.begin() + 1, amplitudes.begin(), amplitudes.end());
}

File: .deprecated/surface-generation/biomes/noise/normal-noise.ts
import { RandomSource } from "../random"
import { PerlinNoise } from "./perlin-noise"

export class NoiseParameters {
    public readonly amplitudes: number[]

    constructor(firstOctave: number, firstAmplitude: number, ...amplitudes: number[])
    constructor(firstOctave: number, amplitudes: number[])
    constructor(
        public readonly firstOctave: number,
        firstAmplitude: number[] | number,
        ...amplitudes: number[]
    ) {
        if (typeof firstAmplitude === "number") {
            this.amplitudes = [firstAmplitude, ...amplitudes]
        } else {
            this.amplitudes = firstAmplitude
        }
    }
}

export class NormalNoise {
    private static readonly INPUT_FACTOR = 1.0181268882175227
    private static readonly TARGET_DEVIATION = 0.3333333333333333
    private readonly valueFactor: number
    private readonly first: PerlinNoise
    private readonly second: PerlinNoise

    public static createLegacyNetherBiome(
        randomSource: RandomSource,
        parameters: NoiseParameters
    ): NormalNoise {
        return new NormalNoise(randomSource, parameters.firstOctave, parameters.amplitudes, false)
    }

    public static create(
        randomSource: RandomSource,
        firstOctave: number,
        ...amplitudes: number[]
    ): NormalNoise {
        return new NormalNoise(randomSource, firstOctave, amplitudes, true)
    }

    public static create2(p_192849_: RandomSource, p_192850_: NoiseParameters): NormalNoise {
        return new NormalNoise(p_192849_, p_192850_.firstOctave, p_192850_.amplitudes, true)
    }

    public static create3(
        randomSource: RandomSource,
        firstOctave: number,
        amplitudes: number[]
    ): NormalNoise {
        return new NormalNoise(randomSource, firstOctave, amplitudes, true)
    }

    private constructor(
        randomSource: RandomSource,
        firstOctave: number,
        amplitudes: number[],
        notLegacy: boolean
    ) {
        if (notLegacy) {
            this.first = PerlinNoise.create2(randomSource, firstOctave, amplitudes)
            this.second = PerlinNoise.create2(randomSource, firstOctave, amplitudes)
        } else {
            this.first = PerlinNoise.createLegacyForLegacyNormalNoise(
                randomSource,
                firstOctave,
                amplitudes
            )
            this.second = PerlinNoise.createLegacyForLegacyNormalNoise(
                randomSource,
                firstOctave,
                amplitudes
            )
        }

        let minAmplitudeIndex = Number.MAX_SAFE_INTEGER
        let maxAmplitudeIndex = Number.MIN_SAFE_INTEGER

        for (let amplitudeIndex = 0; amplitudeIndex < amplitudes.length; amplitudeIndex++) {
            const amplitude = amplitudes[amplitudeIndex]
            if (amplitude != 0.0) {
                minAmplitudeIndex = Math.min(minAmplitudeIndex, amplitudeIndex)
                maxAmplitudeIndex = Math.max(maxAmplitudeIndex, amplitudeIndex)
            }
        }

        this.valueFactor =
            0.16666666666666666 /
            NormalNoise.expectedDeviation(maxAmplitudeIndex - minAmplitudeIndex)
    }

    private static expectedDeviation(v: number): number {
        return 0.1 * (1.0 + 1.0 / (v + 1))
    }

    public getValue(x: number, y: number, z: number): number {
        const inputX = x * NormalNoise.INPUT_FACTOR
        const inputY = y * NormalNoise.INPUT_FACTOR
        const inputZ = z * NormalNoise.INPUT_FACTOR
        return (
            (this.first.getValue(x, y, z) + this.second.getValue(inputX, inputY, inputZ)) *
            this.valueFactor
        )
    }

    public parameters(): NoiseParameters {
        return new NoiseParameters(this.first.firstOctave, this.first.amplitudes)
    }
}
